packages/jaeger-ui/src/components/SearchTracePage/url.test.js

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { getUrl, getUrlState, isSameQuery } from './url';
+
+describe('SearchTracePage/url', () => {
+  const span0 = 'span-0';
+  const span1 = 'span-1';
+  const span2 = 'span-2';
+  const trace0 = 'trace-0';
+  const trace1 = 'trace-1';
+  const trace2 = 'trace-2';
+
+  describe('getUrl', () => {
+    it('handles no args given', () => {
+      expect(getUrl()).toBe('/search');
+    });
+
+    it('handles empty args', () => {
+      expect(getUrl({})).toBe('/search?');
+    });
+
+    it('includes provided args', () => {
+      const paramA = 'aParam';
+      const paramB = 'bParam';
+      expect(getUrl({ paramA, paramB })).toBe(`/search?paramA=${paramA}&paramB=${paramB}`);
+    });
+
+    it('preserves traceID without spanLinks', () => {
+      expect(
+        getUrl({
+          traceID: trace0,
+        })
+      ).toBe(`/search?traceID=${trace0}`);
+    });
+
+    it('converts spanLink and traceID to traceID and span', () => {
+      expect(
+        getUrl({
+          traceID: trace0,
+          spanLinks: {
+            [trace0]: span0,
+          },
+        })
+      ).toBe(`/search?span=${span0}%40${trace0}`);
+    });
+
+    it('handles missing traceID for spanLinks', () => {
+      expect(
+        getUrl({
+          spanLinks: {
+            [trace0]: span0,
+          },
+        })
+      ).toBe(`/search?span=${span0}%40${trace0}`);
+    });
+
+    it('handles empty spanLinks', () => {
+      expect(
+        getUrl({
+          spanLinks: {},
+        })
+      ).toBe(`/search?`);
+    });
+
+    it('converts spanLink and other traceID to traceID and span', () => {
+      expect(
+        getUrl({
+          traceID: [trace0, trace2],
+          spanLinks: {
+            [trace0]: span0,
+          },
+        })
+      ).toBe(`/search?span=${span0}%40${trace0}&traceID=${trace2}`);
```

Comment:
Just FYI, in our use case, you won't find a mix of span@trace1&traceId=${trace2}, it's OK for UI can support that.
We should add a test case for:
```
${span1_1}%20${span1_2}%40${trace1}
&${span2_1}%40${trace2}
&%40${trace3}
```

Reply:
Ah, I didn't realize spanIDs were optional.
I added a check & a test for that.